crates/core/src/error/summary.rs

```diff
@@ -0,0 +1,617 @@
+//! presenting users with a high-level overview.
+
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt::{self, Display, Write},
+    sync::{Arc, Mutex},
+    time::Instant,
+};
+
+use ecow::EcoString;
+
+use crate::error::{ErrorKind, RusticError, RusticResult};
+
+pub type Issues = BTreeMap<IssueCategory, BTreeMap<IssueScope, CondensedIssue>>;
+
+pub type Metrics = BTreeMap<EcoString, EcoString>;
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Default,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    serde::Serialize,
+)]
+pub enum IssueCategory {
+    #[default]
+    Warning,
+    Error,
+    Info,
+}
+
+#[derive(
+    Debug,
+    Clone,
+    Copy,
+    Default,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    derive_more::Display,
+    serde::Serialize,
+)]
+pub enum IssueScope {
+    #[default]
+    Internal,
+    Unknown,
+    UserInput,
+}
+
+#[derive(Debug, Clone, Default, serde::Serialize)]
+pub struct CondensedIssue {
+    /// High-level description of the problem
+    message: EcoString,
+
+    category: IssueCategory,
+
+    /// Number of occurrences
+    count: usize,
+
+    /// Optional diagnostic information, e.g. an error message
+    root_cause: Option<EcoString>,
+}
+
+#[derive(
+    Debug, Clone, Copy, Default, Hash, PartialEq, Eq, derive_more::Display, serde::Serialize,
+)]
+pub enum DisplayOptionKind {
+    #[default]
+    Issues,
+    Timing,
+    Metrics,
+    All,
+}
+
+#[derive(Debug, Clone)]
+pub struct Summary {
+    /// Name of the active context, e.g. a command or operation
+    context: EcoString,
+
+    /// Start time of the collection
+    // Instant cannot be (de-)serialized, for an implementation see:
+    // https://github.com/serde-rs/serde/issues/1375#issuecomment-419688068
+    start_time: Instant,
+
+    /// End time, when the collection is completed
+    // Serialization: See note above
+    end_time: Option<Instant>,
+
+    /// Collection of non-critical warnings   
+    issues: Issues,
+
+    /// Optional custom metrics collected during execution
+    metrics: Metrics,
+
+    /// Display this data
+    display: HashSet<DisplayOptionKind>,
+
+    /// Log enabled
+    log_enabled: bool,
+}
+
+impl Summary {
+    /// Constructor to create an initial empty Summary
+    pub fn new(context: impl Into<EcoString>) -> Self {
+        Self {
+            context: context.into(),
+            start_time: Instant::now(),
+            end_time: None,
+            issues: Issues::default(),
+            metrics: BTreeMap::default(),
+            display: HashSet::from([DisplayOptionKind::default()]),
+            log_enabled: false,
+        }
+    }
+
+    /// Marks the summary as completed, capturing the end time.
+    pub fn complete(&mut self) {
+        self.end_time = Some(Instant::now());
+    }
+
+    pub fn enable_log(&mut self) -> Self {
+        self.log_enabled = true;
+        self.clone()
+    }
```

Comment:
I wonder what is the reason to use clone there?

The builder pattern usually is something like:
```rust
    pub fn enable_log(mut self) -> Self {
        self.log_enabled = true;
        self
    }
```